pkg/nsxt/nsxt_connector.go

```diff
@@ -0,0 +1,155 @@
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/vmware/vsphere-automation-sdk-go/runtime/core"
+	"github.com/vmware/vsphere-automation-sdk-go/runtime/protocol/client"
+	"github.com/vmware/vsphere-automation-sdk-go/runtime/security"
+	"k8s.io/cloud-provider-vsphere/pkg/nsxt/config"
+)
+
+// NewNsxtConnector creates a new NSXT connector
+func NewNsxtConnector(nsxtConfig *config.NsxtConfig) (client.Connector, error) {
+	url := fmt.Sprintf("https://%s", nsxtConfig.Host)
+	securityCtx := core.NewSecurityContextImpl()
+	securityContextNeeded := true
+	if len(nsxtConfig.ClientAuthCertFile) > 0 {
+		securityContextNeeded = false
+	}
+
+	if securityContextNeeded {
+		if len(nsxtConfig.VMCAccessToken) > 0 {
+			if nsxtConfig.VMCAuthHost == "" {
+				return nil, fmt.Errorf("vmc auth host must be provided if auth token is provided")
+			}
+
+			apiToken, err := getAPIToken(nsxtConfig.VMCAuthHost, nsxtConfig.VMCAccessToken)
+			if err != nil {
+				return nil, err
+			}
+
+			securityCtx.SetProperty(security.AUTHENTICATION_SCHEME_ID, security.OAUTH_SCHEME_ID)
+			securityCtx.SetProperty(security.ACCESS_TOKEN, apiToken)
+		} else {
+			if nsxtConfig.User == "" {
+				return nil, fmt.Errorf("username must be provided")
+			}
+
+			if nsxtConfig.Password == "" {
+				return nil, fmt.Errorf("password must be provided")
+			}
+
+			securityCtx.SetProperty(security.AUTHENTICATION_SCHEME_ID, security.USER_PASSWORD_SCHEME_ID)
+			securityCtx.SetProperty(security.USER_KEY, nsxtConfig.User)
+			securityCtx.SetProperty(security.PASSWORD_KEY, nsxtConfig.Password)
+		}
+	}
+
+	tlsConfig, err := getConnectorTLSConfig(nsxtConfig.InsecureFlag, nsxtConfig.ClientAuthCertFile, nsxtConfig.ClientAuthKeyFile, nsxtConfig.CAFile)
+	if err != nil {
+		return nil, err
+	}
+	httpClient := http.Client{
+		Transport: &http.Transport{
+			Proxy:           http.ProxyFromEnvironment,
+			TLSClientConfig: tlsConfig,
+		},
+	}
+
+	connector := client.NewRestConnector(url, httpClient)
+	connector.SetSecurityContext(securityCtx)
+	return connector, nil
+}
+
+func getConnectorTLSConfig(insecure bool, clientCertFile string, clientKeyFile string, caFile string) (*tls.Config, error) {
+	tlsConfig := tls.Config{InsecureSkipVerify: insecure}
+
+	if len(clientCertFile) > 0 {
+		if len(clientKeyFile) == 0 {
+			return nil, fmt.Errorf("Please provide key file for client certificate")
+		}
+
+		cert, err := tls.LoadX509KeyPair(clientCertFile, clientKeyFile)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to load client cert/key pair: %v", err)
+		}
+
+		tlsConfig.Certificates = []tls.Certificate{cert}
+	}
+
+	if len(caFile) > 0 {
+		caCert, err := ioutil.ReadFile(caFile)
+		if err != nil {
+			return nil, err
+		}
+
+		caCertPool := x509.NewCertPool()
+		caCertPool.AppendCertsFromPEM(caCert)
+
+		tlsConfig.RootCAs = caCertPool
+	}
+
+	tlsConfig.BuildNameToCertificate()
+
+	return &tlsConfig, nil
+}
+
+type jwtToken struct {
```

Comment:
@andrewsykim just curious, I saw most of the **struct** definition is on the top of a Golang file - above all functions, is that a coding style? :)
or do we prefer to define the struct right above the function that actually needs it?

Reply:
I think this should be fine, I don't mind struct definitions being between functions